superchain/superchain-version_test.go

```diff
@@ -0,0 +1,109 @@
+package superchain_test
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/ethereum-optimism/superchain-registry/superchain"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+func checkErr(t *testing.T, err error) {
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+// TestContractVersionsCheck will fail if the superchain semver file
+// is not read correctly.
+func TestSemverFile(t *testing.T) {
+	if err := superchain.SuperchainSemver.SanityCheck(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestContractVersionsCheck will check that
+//   - for each declared contract "Foo" in superchain.SuperchainSemver
+//   - for each chain in superchain.OPChains
+//
+// there is a contract address declared for "FooProxy" which has an
+// actual semver matching the declared semver. Actual semvers are
+// read from the L1 chain RPC provider for the chain in question.
+func TestContractVersionsCheck(t *testing.T) {
+
+	semverFields := reflect.VisibleFields(reflect.TypeOf(superchain.SuperchainSemver))
+
+	checkOPChainSatisfiesSemver := func(chain *superchain.ChainConfig) {
+
+		for _, field := range semverFields {
+			contractName := field.Name
+			desiredSemver := reflect.ValueOf(superchain.SuperchainSemver).FieldByName(contractName).String()
+			// ASSUMPTION: we will check the version of the implementation via the declared proxy contract
+			proxyContractName := contractName + "Proxy"
+			rpcEndpoint := superchain.Superchains[chain.Superchain].Config.L1.PublicRPC
+			client, err := ethclient.Dial(rpcEndpoint)
+			checkErr(t, err)
+			r := reflect.ValueOf(superchain.Addresses[chain.ChainID])
+			contractAddressValue := reflect.Indirect(r).FieldByName(proxyContractName)
+			if contractAddressValue == (reflect.Value{}) {
+				t.Errorf("%10s:%-20s:%-30s has version UNSPECIFIED, but should have version %s", chain.Superchain, chain.Name, proxyContractName, desiredSemver)
+				continue
+			}
+			actualSemver, err := getVersionWithRetries(context.Background(), common.Address(contractAddressValue.Bytes()), client)
+			if err != nil {
+				t.Errorf("RPC endpoint %s: %s", rpcEndpoint, err)
+			}
+			if desiredSemver != actualSemver {
+				t.Errorf("%10s:%-20s:%-30s has version %s but should have version %s", chain.Superchain, chain.Name, contractName, actualSemver, desiredSemver)
+			}
+		}
+	}
+
+	for _, chain := range superchain.OPChains {
+		if strings.ToLower(chain.Superchain) != "mainnet" {
+			continue // concentrate on mainnet superchain for now
```

Comment:
Flagging this. I intend for this to be a temporary thing while we work some other questions out. 
```suggestion
			continue // TODO Check all superchains
```